packages/expo-updates/android/src/main/java/expo/modules/updates/loader/ExpoMultipartStreamReader.kt

```diff
@@ -0,0 +1,162 @@
+import okio.BufferedSource
+import okio.ByteString
+import java.io.IOException
+import java.util.HashMap
+import kotlin.math.max
+
+/**
+ * Fork of {@link com.facebook.react.devsupport.MultipartStreamReader} that doesn't necessarily
+ * expect a preamble (first boundary is not necessarily preceded by CRLF).
+ */
+class ExpoMultipartStreamReader(private val source: BufferedSource, private val boundary: String) {
+  private var mLastProgressEvent: Long = 0
+
+  interface ChunkListener {
+    /** Invoked when a chunk of a multipart response is fully downloaded.  */
+    @Throws(IOException::class)
+    fun onChunkComplete(headers: Map<String, String>?, body: Buffer, isLastChunk: Boolean)
+
+    /** Invoked as bytes of the current chunk are read.  */
+    @Throws(IOException::class)
+    fun onChunkProgress(headers: Map<String, String>, loaded: Long, total: Long)
+  }
+
+  private fun parseHeaders(data: Buffer): Map<String, String> {
+    val headers: MutableMap<String, String> = HashMap()
+    val text = data.readUtf8()
+    val lines = text.split(CRLF.toRegex()).toTypedArray()
+    for (line in lines) {
+      val indexOfSeparator = line.indexOf(":")
+      if (indexOfSeparator == -1) {
+        continue
+      }
+      val key = line.substring(0, indexOfSeparator).trim { it <= ' ' }
+      val value = line.substring(indexOfSeparator + 1).trim { it <= ' ' }
+      headers[key] = value
+    }
+    return headers
+  }
+
+  @Throws(IOException::class)
+  private fun emitChunk(chunk: Buffer, done: Boolean, listener: ChunkListener) {
+    val marker = ByteString.encodeUtf8(CRLF + CRLF)
+    val indexOfMarker = chunk.indexOf(marker)
+    if (indexOfMarker == -1L) {
+      listener.onChunkComplete(null, chunk, done)
+    } else {
+      val headers = Buffer()
+      val body = Buffer()
+      chunk.read(headers, indexOfMarker)
+      chunk.skip(marker.size().toLong())
+      chunk.readAll(body)
+      listener.onChunkComplete(parseHeaders(headers), body, done)
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun emitProgress(
+    headers: Map<String, String>?,
+    contentLength: Long,
+    isFinal: Boolean,
+    listener: ChunkListener?
+  ) {
+    if (headers == null || listener == null) {
+      return
+    }
+    val currentTime = System.currentTimeMillis()
+    if (currentTime - mLastProgressEvent > 16 || isFinal) {
+      mLastProgressEvent = currentTime
+      val headersContentLength = if (headers["Content-Length"] != null) headers["Content-Length"]!!
+        .toLong() else 0
+      listener.onChunkProgress(headers, contentLength, headersContentLength)
+    }
+  }
+
+  /**
+   * Reads all parts of the multipart response and execute the listener for each chunk received.
+   *
+   * @param listener Listener invoked when chunks are received.
+   * @return If the read was successful
+   */
+  @Throws(IOException::class)
+  fun readAllParts(listener: ChunkListener): Boolean {
+    val delimiter = ByteString.encodeUtf8("--$boundary$CRLF")
```

Comment:
So am I reading it correctly that other than a kotlin conversion, this is the important difference from https://github.com/facebook/react-native/blob/main/ReactAndroid/src/main/java/com/facebook/react/devsupport/MultipartStreamReader.java ?

Reply:
Correct. I guess the multipart response their reader is expecting is a very particular one that always contains a preamble. Our's does not necessarily contain one.